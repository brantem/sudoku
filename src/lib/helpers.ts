import { twMerge } from 'tailwind-merge';
import clsx, { ClassValue } from 'clsx';

export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

export const getTimeDiff = (a: number, b: number) => {
  const v = b - a;
  return {
    hours: Math.floor(v / (1000 * 60 * 60)),
    minutes: Math.floor((v % (1000 * 60 * 60)) / (1000 * 60)),
    seconds: Math.floor((v % (1000 * 60)) / 1000),
  };
};

export const padStart = (s: number, maxLength = 2, fillString = '0') => {
  return String.prototype.padStart.call(s, maxLength, fillString);
};

export const shuffle = <T>(a: T[]): T[] => {
  const b = [...a];
  for (let i = b.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [b[i], b[j]] = [b[j], b[i]];
  }
  return b;
};
