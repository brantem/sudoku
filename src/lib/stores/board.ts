export const useBoardStore = () => ({
  board: [
    [
      [4, -4],
      [1, 1],
      [2, 2],
      [3, -3],
      [6, -6],
      [5, -5],
      [7, 7],
      [8, 8],
      [9, 9],
    ],
    [
      [7, -7],
      [5, -5],
      [6, 6],
      [9, 9],
      [2, -2],
      [8, -8],
      [3, -3],
      [4, -4],
      [1, 1],
    ],
    [
      [3, 3],
      [9, 9],
      [8, -8],
      [4, 4],
      [1, -1],
      [7, -7],
      [5, 5],
      [2, -2],
      [6, 6],
    ],
    [
      [9, -9],
      [8, -8],
      [5, -5],
      [6, -6],
      [7, -7],
      [1, 1],
      [2, 2],
      [3, -3],
      [4, -4],
    ],
    [
      [6, -6],
      [2, -2],
      [7, 7],
      [5, -5],
      [3, -3],
      [4, 4],
      [9, 9],
      [1, -1],
      [8, 8],
    ],
    [
      [1, -1],
      [3, 3],
      [4, 4],
      [8, -8],
      [9, -9],
      [2, -2],
      [6, -6],
      [5, 5],
      [7, 7],
    ],
    [
      [5, 5],
      [7, 7],
      [1, -1],
      [2, -2],
      [4, -4],
      [6, 6],
      [8, 8],
      [9, -9],
      [3, 3],
    ],
    [
      [8, -8],
      [4, -4],
      [9, -9],
      [7, -7],
      [5, 5],
      [3, -3],
      [1, -1],
      [6, -6],
      [2, -2],
    ],
    [
      [2, 2],
      [6, 6],
      [3, -3],
      [1, -1],
      [8, 8],
      [9, -9],
      [4, 4],
      [7, -7],
      [5, -5],
    ],
  ],
});
