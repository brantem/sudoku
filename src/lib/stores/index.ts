export * from 'lib/stores/board';
