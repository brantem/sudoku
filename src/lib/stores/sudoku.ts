import { useState, useEffect } from 'react';
import { useStore } from 'zustand';
import { createStore } from 'zustand/vanilla';
import { persist } from 'zustand/middleware';

import { Difficulty, Coord, Board } from 'types/sudoku';
import { shuffle } from 'lib/helpers';

type SudokuState = {
  startedAt: number | null;
  completedAt: number | null;
  difficulty: Difficulty;
  setDifficulty: (difficulty: number) => void;

  board: Board;
  values: Record<number, number>;
  generate: () => void;
  solve: () => void;
  filled: number;

  coord: Coord | null;
  setCoord: (coord: Coord | null) => void;
  updateCell: (value: number) => void;
};

const isCellValid = (board: Board, x: number, y: number, num: number) => {
  for (let i = 0; i < 9; i++) {
    if (board[x][i][0] === num) return false;
  }

  for (let i = 0; i < 9; i++) {
    if (board[i][y][0] === num) return false;
  }

  const sx = Math.floor(x / 3) * 3;
  const sy = Math.floor(y / 3) * 3;
  for (let x = sx; x < sx + 3; x++) {
    for (let y = sy; y < sy + 3; y++) {
      if (board[x][y][0] === num) return false;
    }
  }

  return true;
};

const fillCell = (board: Board, x: number, y: number) => {
  const nums = shuffle([1, 2, 3, 4, 5, 6, 7, 8, 9]);
  for (const num of nums) {
    if (!isCellValid(board, x, y, num)) continue;
    board[x][y] = [num, 0];

    if (x === 8 && y === 8) return;

    if (y === 8) {
      fillCell(board, x + 1, 0);
    } else {
      fillCell(board, x, y + 1);
    }

    if (board[8][8][0] !== 0) return;
    board[x][y] = [0, 0];
  }
};

export const sudokuStore = createStore<SudokuState>()(
  persist(
    (set, get) => ({
      startedAt: null,
      completedAt: null,
      difficulty: Difficulty.Easy,
      setDifficulty: (difficulty) => {
        set({ difficulty });
        get().generate();
      },

      board: [],
      values: {},
      generate: () => {
        const board: Board = [];
        let coords = [];
        for (let x = 0; x < 9; x++) {
          board[x] = [];
          for (let y = 0; y < 9; y++) {
            board[x][y] = [0, 0];
            coords.push([x, y]);
          }
        }
        coords = shuffle(coords);

        for (let i = 0; i < 9; i += 3) {
          fillCell(board, i, i);
        }

        const values: Record<number, number> = {};
        let filled = 0;
        for (let i = 0; i < get().difficulty; i++) {
          const [x, y] = coords[i];
          board[x][y][1] = board[x][y][0] * -1;
          values[board[x][y][0]] = (values[board[x][y][0]] || 0) + 1;
          filled++;
        }

        set({ startedAt: Date.now(), completedAt: null, board, values, coord: null, filled });
      },
      solve: () => {
        const { board } = get();
        for (let x = 0; x < 9; x++) {
          for (let y = 0; y < 9; y++) {
            if (board[x][y][1] < 0) continue;
            board[x][y][1] = board[x][y][0];
          }
        }
        set({
          completedAt: Date.now(),
          board,
          values: { 1: 9, 2: 9, 3: 9, 4: 9, 5: 9, 6: 9, 7: 9, 8: 9, 9: 9 },
          coord: null,
          filled: 81,
        });
      },
      filled: 0,

      coord: null,
      setCoord: (coord) => set({ coord }),
      updateCell: (value) => {
        const { board, values, generate, filled, coord } = get();
        if (!coord) return;

        const [x, y] = coord;
        if (board[x][y][1] < 0 || board[x][y][0] === board[x][y][1]) return;

        if (board[x][y][1] !== 0) values[board[x][y][1]]--;
        if (value !== 0) values[value] = (values[value] || 0) + 1;

        board[x][y][1] = value;
        const _filled = value === board[x][y][0] ? filled + 1 : filled;
        set({ board, values, filled: _filled, completedAt: _filled === 81 ? Date.now() : null });
        if (_filled === 81) setTimeout(generate, 1000);
      },
    }),
    { name: 'sudoku' },
  ),
);

/* c8 ignore start */
const dummy = {
  startedAt: null,
  completedAt: null,
  difficulty: Difficulty.Easy,
  setDifficulty: () => {},

  board: [],
  values: {},
  generate: () => {},
  solve: () => {},
  filled: 0,

  coord: null,
  setCoord: () => {},
  updateCell: () => {},

  isCellValid: () => true,
};

// https://github.com/pmndrs/zustand/issues/1145
export function useSudokuStore(): SudokuState;
export function useSudokuStore<T>(selector: (state: SudokuState) => T, equals?: (a: T, b: T) => boolean): T;
export function useSudokuStore(selector?: any, equals?: any) {
  const store = useStore(sudokuStore, selector, equals);
  const [isHydrated, setHydrated] = useState(false);
  useEffect(() => setHydrated(true), []);
  return isHydrated ? store : selector ? selector(dummy) : dummy;
}
/* c8 ignore stop */
