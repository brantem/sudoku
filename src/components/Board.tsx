import { cn } from 'lib/helpers';
import { useSudokuStore } from 'lib/stores';

const Board = () => {
  const { board, coord, setCoord } = useSudokuStore((state) => ({
    board: state.board,
    coord: state.coord,
    setCoord: state.setCoord,
  }));
  const size = board.length || 9;

  return (
    <div className="flex flex-col overflow-hidden rounded-xl border-[3px] border-neutral-500">
      {[...new Array(size)].map((_, x) => (
        <div key={x} className={cn('flex', x > 0 && x % 3 === 0 && 'border-t-[3px] border-neutral-500')}>
          {[...new Array(size)].map((_, y) => {
            const cell = board.length ? board[x][y] : [0, 0];
            const value = cell[1] < 0 ? cell[0] : cell[1] === 0 ? '' : cell[1];

            const isFilled = cell[1] + cell[0] === 0;
            const isActive = coord ? coord[0] === x && coord[1] === y : false;

            const activeCell = coord && board[coord[0]][coord[1]];
            const shouldActive = activeCell
              ? activeCell[1] !== 0 && activeCell[activeCell[1] > 0 ? 1 : 0] === value
              : false;

            return (
              <button
                key={y}
                className={cn(
                  'flex h-[42px] flex-1 items-center justify-center border-neutral-500 bg-white text-2xl font-semibold sm:h-14 sm:text-3xl',
                  'dark:bg-neutral-800',
                  y > 0 && 'border-l',
                  x % 3 !== 0 && 'border-t',
                  y > 0 && y % 3 === 0 && 'border-l-[3px]',
                  (isFilled || value === cell[0]) && 'bg-neutral-200 dark:bg-neutral-900',
                  shouldActive || isActive
                    ? 'bg-blue-400 dark:bg-blue-700'
                    : 'hover:bg-blue-200 dark:hover:bg-blue-900',
                )}
                data-active={shouldActive || isActive}
                onClick={() => setCoord([x, y])}
              >
                {value}
              </button>
            );
          })}
        </div>
      ))}
    </div>
  );
};

export default Board;
