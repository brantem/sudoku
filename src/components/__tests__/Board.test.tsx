import { render, screen, act, renderHook } from '@testing-library/react';
import '@testing-library/jest-dom';

import Board from 'components/Board';

import { boardStore, useBoardStore } from 'lib/stores';

describe('Board', () => {
  afterEach(() => {
    act(() =>
      boardStore.setState({
        board: [
          [
            [1, -1],
            [2, 2],
            [3, -3],
          ],
          [
            [4, 0],
            [5, -5],
            [6, 0],
          ],
          [
            [7, -7],
            [8, 0],
            [9, -9],
          ],
        ],
        coord: null,
      }),
    );
  });

  it('render successfully', () => {
    const { container } = render(<Board />);

    expect(container).toMatchSnapshot();
  });

  it('should set coord', () => {
    const { result } = renderHook(() => useBoardStore());
    const setCoord = vi.spyOn(result.current, 'setCoord').mockImplementationOnce(() => {});

    render(<Board />);

    act(() => screen.getByText('1').click());
    expect(setCoord).toHaveBeenCalledWith([0, 0]);
  });

  it('should activate other cells with the same value', () => {
    act(() =>
      boardStore.setState({
        board: [
          [
            [1, -1],
            [2, 1],
          ],
          [
            [3, -3],
            [4, -4],
          ],
        ],
      }),
    );
    const { result } = renderHook(() => useBoardStore());

    render(<Board />);

    screen.getAllByText('1').forEach((el) => expect(el).toHaveAttribute('data-active', 'false'));
    act(() => result.current.setCoord([0, 0]));
    screen.getAllByText('1').forEach((el) => expect(el).toHaveAttribute('data-active', 'true'));
    act(() => result.current.setCoord([0, 1]));
    screen.getAllByText('1').forEach((el) => expect(el).toHaveAttribute('data-active', 'true'));
  });
});
