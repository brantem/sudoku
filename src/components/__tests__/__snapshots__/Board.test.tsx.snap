// Vitest Snapshot v1

exports[`Board > render successfully 1`] = `
<div>
  <div
    class="flex flex-col overflow-hidden rounded-xl border-[3px] border-neutral-500"
  >
    <div
      class="flex"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex border-t-2 border-neutral-400"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex border-t-2 border-neutral-400"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
    <div
      class="flex"
    >
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-t border-l-2 bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
      <button
        class="flex h-9 flex-1 items-center justify-center border-neutral-400 text-2xl font-semibold sm:h-12 sm:text-3xl border-l border-t bg-neutral-200 hover:bg-neutral-200/50"
        data-active="false"
      />
    </div>
  </div>
</div>
`;
